### PR TITLE
Occassional issues with watched status (duration set to 0)

### PR DESCRIPTION
**Pull request by wagnerch**
### Check if this PR fulfills these requirements:
* [X] My changes respect the [Kodi add-on development rules](https://kodi.wiki/view/Add-on_rules)
* [X] I have read the [**CONTRIBUTING**](Contributing.md) document.
* [X] I made sure there wasn't another one [Pull Requests](../../pulls) opened for the same update/change
* [X] I have successfully tested my changes locally

#### Types of changes
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Feature change (non-breaking change which change behaviour of an existing functionality)
- [ ] Improvement (non-breaking change which improve functionality)
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Description

It appears if you let a video play to completion when Kodi is stopping the player there is a chance the on_tick fires pulls the player state and the time is set to 0.

~The solution is a bit of a hack, your choice if there is a better way.  But the fix seems to work.  The basic idea is to only update last_player_state (which is what is used when sending the stop event) if the elapsed_seconds has increased.  To accommodate rewinds I also force an update of the last_player_state in the on_playback_seek method.~

~Also in on_tick, I moved the saving of player state because when the video has been paused for more than 30 minutes it sends last_player_state so I didn't think it made sense to update it, but I may not understand the consequences of doing that.~

### CODE REVIEW

**Comment by wagnerch** [comment]
I may have found a slightly better way, going to rework this commit.  It appears totaltime is also zero'd out when this occurs.

2020-05-09 17:57:39.277 T:5668   ERROR: [plugin.video.netflix (0)] current player state: {u'totaltime': {u'hours': 0, u'seconds': 0, u'minutes': 0, u'milliseconds': 0}, u'currentaudiostream': {u'index': 2, u'name': u'DD+ ', u'language': u'eng', u'channels': 6, u'codec': u'eac3', u'bitrate': 655360}, u'currentvideostream': {u'index': 0, u'name': u'', u'language': u'unk', u'height': 720, u'width': 1280, u'codec': u'vp9'}, u'elapsed_seconds': 0, u'time': {u'hours': 0, u'seconds': 0, u'minutes': 0, u'milliseconds': 0}, u'subtitleenabled': True, u'percentage': 0.0, u'currentsubtitle': {u'index': 1, u'name': u'', u'language': u'eng'}, u'audiostreams': [{u'index': 0, u'name': u'AAC ', u'language': u'ger', u'channels': 2, u'codec': u'aac', u'bitrate': 131072}, {u'index': 1, u'name': u'DD+ ', u'language': u'ger', u'channels': 6, u'codec': u'eac3', u'bitrate': 655360}, {u'index': 2, u'name': u'DD+ ', u'language': u'eng', u'channels': 6, u'codec': u'eac3', u'bitrate': 655360}, {u'index': 3, u'name': u'DD+ ', u'language': u'eng', u'channels': 2, u'codec': u'eac3', u'bitrate': 131072}, {u'index': 4, u'name': u'DD+ ', u'language': u'eng', u'channels': 2, u'codec': u'eac3', u'bitrate': 131072}, {u'index': 5, u'name': u'DD+ ', u'language': u'eng', u'channels': 6, u'codec': u'eac3', u'bitrate': 655360}, {u'index': 6, u'name': u'AAC ', u'language': u'spa', u'channels': 2, u'codec': u'aac', u'bitrate': 131072}, {u'index': 7, u'name': u'DD+ ', u'language': u'spa', u'channels': 6, u'codec': u'eac3', u'bitrate': 655360}, {u'index': 8, u'name': u'AAC ', u'language': u'spa', u'channels': 2, u'codec': u'aac', u'bitrate': 131072}, {u'index': 9, u'name': u'DD+ ', u'language': u'spa', u'channels': 6, u'codec': u'eac3', u'bitrate': 655360}, {u'index': 10, u'name': u'DD+ ', u'language': u'fre', u'channels': 2, u'codec': u'eac3', u'bitrate': 131072}, {u'index': 11, u'name': u'DD+ ', u'language': u'fre', u'channels': 6, u'codec': u'eac3', u'bitrate': 655360}, {u'index': 12, u'name': u'AAC ', u'language': u'ita', u'channels': 2, u'codec': u'aac', u'bitrate': 131072}, {u'index': 13, u'name': u'DD+ ', u'language': u'ita', u'channels': 6, u'codec': u'eac3', u'bitrate': 655360}], u'subtitles': [{u'index': 0, u'name': u'', u'language': u'ger'}, {u'index': 1, u'name': u'', u'language': u'eng'}, {u'index': 2, u'name': u'', u'language': u'eng'}, {u'index': 3, u'name': u'', u'language': u'eng'}, {u'index': 4, u'name': u'', u'language': u'spa'}, {u'index': 5, u'name': u'', u'language': u'spa'}, {u'index': 6, u'name': u'', u'language': u'fre'}, {u'index': 7, u'name': u'', u'language': u'fre'}, {u'index': 8, u'name': u'', u'language': u'ita'}, {u'index': 9, u'name': u'', u'language': u'chi'}, {u'index': 10, u'name': u'', u'language': u'chi'}]}
2020-05-09 17:57:39.277 T:5668 WARNING: [plugin.video.netflix (0)] AMVideoEvents: not saving player state due to unexpected playback position

**Comment by CastagnaIT** [comment]
> It appears if you let a video play to completion when Kodi is stopping the player there is a chance the on_tick fires pulls the player state and the time is set to 0.

another kodi bug that shit..
in what Kodi version?

> Also in on_tick, I moved the saving of player state because when the video has been paused for more than 30 minutes it sends last_player_state so I didn't think it made sense to update it, but I may not understand the consequences of doing that.

this all right
I have made many changes to this module to achieve an acceptable solution to develop this
It's an oversight but it doesn't affect anything

I do not understand because you doubled an existing variable on line 166 is always the same value

I need to take a closer look i will let you know later

**Comment by CastagnaIT** [comment]
I also need a full debug log, with the error you attach before

**Comment by CastagnaIT** [comment]
well, i think to understand
when a user Stop the playback, the stop notification (fired by Kodi moditor) arrives delayed, so is not immediate, in the meantime of this interval of time (due to delay) the nf service keep reporting tick calls with incorrect player_state content, because actually the playback is already stopped and kodi api returns partial data.

then i think this problem can be solved differently
in the action_controller there is already a similar fix that cause the stop of service tick calls: https://github.com/CastagnaIT/plugin.video.netflix/blob/master/resources/lib/services/playback/action_controller.py#L155 
https://github.com/CastagnaIT/plugin.video.netflix/blob/master/resources/lib/services/playback/action_controller.py#L91

then i think the best way is add a the new case directly in the action_controller.
So add a new protected variable `self._last_player_state` and before the return of `_get_player_state()` add:

```python
if self._last_player_state and (not player_state['audiostreams'] or not player_state['elapsed_seconds']):
    player_state = {}
return player_state
```

in this way the fix cover all modules not only a single action manager module
and no more service ticks calls with partial data

**Comment by wagnerch** [comment]
> another kodi bug that shit..
> in what Kodi version?

Ahh, sorry.  Yes it's Kodi 18.6

**Comment by wagnerch** [comment]
> then i think the best way is add a the new case directly in the action_controller.
> So add a new protected variable `self._last_player_state` and before the return of `_get_player_state()` add:
> 
> ```python
> if self._last_player_state and (not player_state['audiostreams'] or not player_state['elapsed_seconds']):
>     player_state = {}
> else:
>     self._last_player_state = player_state
> return player_state
> ```
> 
> in this way the fix cover all modules not only a single action manager module
> and no more service ticks calls with partial data

Yeah, that is better, however I am not sure if self._last_player_state condition in the if is necessary, if we leave that in then if the audiostreams is empty (the original case you were trying to cover) might be missed if the first tick ends up with this condition.  Is that likely to happen?  If we leave it out then if on the first tick the return would be the same as before if we initialize _last_player_state to {}.

It seems like the next logical step is to remove the last_player_state logic from am_video_events.py and perhaps add the player_state to on_playback_stopped?  I would hope that if the video is paused >= 30m that the player state is still consistent, or do you think that logic in am_video_events.py is still necessary?

**Comment by wagnerch** [comment]
By the way, here is a log of the issue with the fix from https://github.com/CastagnaIT/plugin.video.netflix/commit/38dacef1d5603814db379d1cf01c6c658fb48fbc

[kodi.log](https://github.com/CastagnaIT/plugin.video.netflix/files/4605785/kodi.log)

**Comment by CastagnaIT** [comment]
One thing is true my operator "or" in this way it's wrong
but also your solution is wrong because does not take into account a possible value of 0 in elapsed_seconds.
sidenote add multiple return where is not necessary is a wrong implemetation

more complex but this can be a solution:
```python
if (player_state['audiostreams'] and player_state['elapsed_seconds']) or (player_state['audiostreams'] and not player_state['elapsed_seconds'] and not self._last_player_state):
    self._last_player_state = player_state
else:
    player_state = self._last_player_state

return player_state
```

this taken in account also the possibility that the elapsed_seconds starts from "0" when starts playback

now _last_player_state is set only if all is right
even if there are problems at the first start return always self._last_player_state that in initialize_playback will be cleaned

---

> It seems like the next logical step is to remove the last_player_state logic from am_video_events.py and perhaps add the player_state to on_playback_stopped? I would hope that if the video is paused >= 30m that the player state is still consistent, or do you think that logic in am_video_events.py is still necessary?

yes is possible to add the player_state to on_playback_stopped
and remove self.last_player_state in am_video_events.py
and https://github.com/CastagnaIT/plugin.video.netflix/blob/master/resources/lib/services/playback/am_video_events.py#L55
the self.last_player_state can be changed with player_state of function param

the pause check of 30m still needed, is a limit imposed for not continuously send event request data to netflix when user leave the device in pause for long time,
this in the website is done automatically with the user session but not here

**Comment by wagnerch** [comment]
here are the other 2 commits to add player_state to on_playback_stopped, using self._last_player_state, and the other is to remove last_player_state from am_video_events.py

This should also fix an issue with seeking to the end of the video, one problem I noticed is because the watched status would sometimes be 0 if I opened the video again and tried to seek to the end to push the watch status again it would often have maybe a book mark of maybe 5 minutes in.

File: resources/lib/services/playback/action_controller.py [hunk 1]
```diff
@@ -149,18 +152,23 @@ def _get_player_state(self, player_id=None):
...
         # convert time dict to elapsed seconds
         player_state['elapsed_seconds'] = (
             player_state['time']['hours'] * 3600 +
             player_state['time']['minutes'] * 60 +
             player_state['time']['seconds'])
 
+        # Sometime may happen that when you stop playback, a player status without data is read,
+        # or while Kodi is stopping video it will return time as 0
+        if (player_state['audiostreams'] and player_state['elapsed_seconds']) or \
+           (player_state['audiostreams'] and not player_state['elapsed_seconds'] and not self._last_player_state):
```
Comment (CastagnaIT) [inline resources/lib/services/playback/action_controller.py:164]: to split if condition, enclose all if condition inside parenthesis (..condition..) and remove slash

File: resources/lib/services/playback/action_controller.py [hunk 2]
```diff
@@ -53,6 +54,7 @@ def initialize_playback(self, data):
             AMVideoEvents(),
             AMUpNextNotifier()
         ]
+        self._last_player_state = {}
```
[the remark this answers is not in the document]
Comment (CastagnaIT) [reply resources/lib/services/playback/action_controller.py:57]: can also stand above the others as first

**Comment by CastagnaIT** [comment]
do not close, use always same PR

**Comment by wagnerch** [comment]
> do not close, use always same PR

Was trying to get crafty on pushing the rework, didn't work out and Github auto-closed it.  Pushed the changes now.

**Comment by CastagnaIT** [comment]
now is perfect, but you have to fix a pylint issue
see CI details

**Comment by CastagnaIT** [comment]
another thing missing to change on_playback_stopped inherit of am_section_skipping.py
if you want you can do it also to separate commit

**Comment by wagnerch** [comment]
> 
> 
> another thing missing to change on_playback_stopped inherit of am_section_skipping.py
> if you want you can do it also to separate commit

I think it's covered in this commit or did I miss something else?

https://github.com/CastagnaIT/plugin.video.netflix/blob/7b400335d5b03b55991a059613559e265a7d24a1/resources/lib/services/playback/am_section_skipping.py#L86

**Comment by CastagnaIT** [comment]
oh yes sorry, see the changes on github after a force push, github do not update the page, show previous code and add a Refresh button.... LOL...
miss to click Refresh button

**Pull request merged**